Hotfix/fix dcs get results

itayx opened the pull request:
**To be merged after #807**

all DCS code has been ported #562 
DCS Validation has been started on both RDKB/UGW platforms.

This PR is used to fix an issue with the DCS feature's "get results" flow
This PR contains:
- a fix to the bml message results_size variable
- changes to the get results flow in the son_management
- extended logs in bml_internal
- fix casting error in bml_internal
- fix validation error in bml_internal

itayx committed:
controller: bml: fix results type

To receive the channel scan results the caller sends an empty initialized
results list, the allocated storage is then used to store the incoming
results, previously this flow caused a `segmentation fault` error and lead
to a crush scenario.

To fix the issue the results variable in the caller function needed to
be changed from `BML_NEIGHBOR_AP**` to `BML_NEIGHBOR_AP*` since the
initialization moved from within the bwl_internal to the caller
function.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
yaml: Change type of results_size

The ACTION_BML_CHANNEL_SCAN_GET_RESULTS_RESPONSE message uses a
parameter called results_size which was set to `uint8_t`.
This caused an error where durin the message's class_swap function it
crashes causing the controller process to crush as well
since after the swap it is immidiatly used as the limiter for a loop

To fix the issue all that was needed was to change the results_size from
a `uint32_t` to a `uint8_t` since values set to `uint8_t` are not
effected by the swap

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
yaml: Change type of results_size AutoGenerated

The ACTION_BML_CHANNEL_SCAN_GET_RESULTS_RESPONSE message uses a
parameter called results_size which was set to `uint8_t`.
This caused an error where durin the message's class_swap function it
crashes causing the controller process to crush as well
since after the swap it is immidiatly used as the limiter for a loop

To fix the issue all that was needed was to change the results_size from
a `uint32_t` to a `uint8_t` since values set to `uint8_t` are not
effected by the swap

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: son_managment: fix results allocation

The Dynamic channel selection feature is used to get the neightboring AP
using a channel scan, after the scan is done the results are stored to
the controller's DB and are pulled in the son_managment when a BML
message is sent, The son_managment then sends those results back to
the BML. The son_managment can send as meny CMDUs as needed to sent all
the available results for a given radio mac.

This flow was not optimised and could even result in a crush if the
results exeeded the CMDU's capacity. to solve the issue the
allocation process is not used for each result dump that is being
sent, and not for a predetermined size.

Each error case now uses it's own generated response instead of
using a single response.

To reduce duplicate code, add lambda functions for "send cmdu" and
"create new cmdu".

Signed-off-by: alex kanter <alexx.kanter@intel.com>
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml_internal: extend logs

Extend the logs in the bml_internal's 'get results' flow for the
'get results' flow to allow for a easier to understand debug output.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: bml_internal: fix validation

The bml_internal is called with a pointer to the pre-allocated
results array, a max size constant and two references to the
outgoing size & status.

since the code uses references and not pointers, the validation
needs to be changed accordingly, and add validation to the max size
constant.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

tomereli reviewed (approved):
Most comments are easy to fix, so approving. Great split to commits and good commit messages!

Commit:
bwl: dwpal: fix signal strength value

signal strength value is signed, but we are using netlink attribute that returnes unsigned value.

need to cast to signed int since there is netlink attribute that returnes signed vaue.
also, removed /100 in case of scaled value.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml_defs: fix neighbor ap structure

bssid copied from CMDU message so it has to be the same
in both sChannelScanResults/BML_NEIGHBOR_AP structures.
Changed to uint8.
Also, parameter descriptions should be above the parameter.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
cli_bml: int array print helper function fix

string_from_int_array works only with 32int arrays.
It cannot handle 8/16int arrays since it gets
void pointer to the array but it doesn’t know how many bytes per object.

currently used for 8/32int arrays so moved to header
and implemented with 2 type of signatures.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
cli_bml: fix dcs scan results print message

some fields in scan results print message not shown correctly.

need to add int casting to enums.
print bssid in mac format.
Also, add units in parameter name.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2020,12 +2020,12 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
     unsigned int results_count = max_results_size;
     int ret                    = bml_get_dcs_scan_results(ctx, radio_mac.c_str(),
-                                       reinterpret_cast<BML_NEIGHBOR_AP **>(&results),
+                                       reinterpret_cast<BML_NEIGHBOR_AP *>(results),
```
The reinterpret_cast can be removed now I think. It's due to this cast that we didn't notice the issue before.

itayx replied:
Changed to `static_cast`, since the type of `results` is `BML_NEIGHBOR_AP[max_results_size]` while the requested type is `BML_NEIGHBOR_AP*`. And while the two are identical, I believe it is better to cast the type ourselves

arnout replied:
No, the two really are identical. Casting would give the wrong impression.

arnout replied:
Let me qualify that... By adding an explicit static cast, you hide a possible problem where the type of either `bml_get_dcs_scan_results` or `results` changes. True, the number of casts that are allowed with static_cast is *very* limited, but there are still a few possible, and it is very unlikely that they are what we want there. So removing the static cast may avoid bugs in the future (or rather, make sure they are captured by the compiler).

itayx replied:
Removed the cast all together
It builds with no issues, before when we used `BML_NEIGHBOR_AP **` without a cast the build failed, but when we changed the API call to `BML_NEIGHBOR_AP *` we don't need an explicit cast

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -70,7 +70,8 @@ static void config_logger(const std::string log_file = std::string())
 {
-    string_utils::copy_string(res_out.ap_SSID, res_in.ssid,
+    string_utils::copy_string(reinterpret_cast<char *>(res_out.ap_SSID),
+                              reinterpret_cast<const char *>(res_in.ssid),
```
This is (again) a bogus change that does not solve the issue that is claimed in the commit message.

res_out.ap_SSID is already of type char*, so adding a cast does nothing.

res_in.ssid is also a char*, so it's essentially just a const_cast, and unnecessary because the const is added.

Also, if a cast would be required, it should be a static_cast, not a reinterpret_cast, because the latter is simply a C-style cast in another jacket.

itayx replied:
You're right, the segmentation error solved in an earlier commit, it's an oversight that this commit was included so I'm removing the entire commit

arnout commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1574,7 +1574,7 @@ int bml_internal::get_dcs_scan_results(const sMacAddr &mac, BML_NEIGHBOR_AP *res
                                        bool is_single_scan)
 {
-    if (!results || !results_size || !result_status) {
+    if (!results || max_results_size == 0) {
```
Actually, the `!results` bit is also not entirely correct: that is a check of the input parameter validity (similar to the `!ctx` check in bml_get_dcs_scan_results).

So better would be to have a separate check for results and for max_results_size.

(nitpick)

itayx replied:
I understand and agree with your point that `!results` is an  input parameter validity check
But I think that `max_results_size == 0` could also be classified as a parameter validity check since it check the validity of the parameter `max_results_size`

Since the parameter `max_results_size` is defined as a `const unsigned int` is cannot have a signed value, but can still be set as 0, which is invalid for the "get results" flow since it would mean that the client has requested 0 results
therefore the error `"Function is called, but no data is being requested!"`

arnout replied:
OK fair enough.

arnout commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -382,9 +382,13 @@ static bool translate_nl_data_to_bwl_results(sChannelScanResults &results,
-        results.signal_strength_dBm = (nla_get_u8(bss[NL80211_BSS_SIGNAL_UNSPEC])) / 100;
+        //signal strength of the probe response/beacon in unspecified units, scaled to 0..100 <u8>
+        results.signal_strength_dBm =
+            static_cast<int32_t>(nla_get_u8(bss[NL80211_BSS_SIGNAL_UNSPEC]));
```
The static cast does nothing here, it is implied by the type change from u8 to int32.

Also, if the value is in the range 0..100, it means we can get something up to 100dBm... Is that a sane range? It's unspecified anyway so it probably doesn't matter, but still...

kantera800 replied:
I wasn't sure about it either but this is what the nl80211 api provides.
https://lxr.openwrt.org/source/iwinfo/api/nl80211.h#L3407

never saw this type in the logs but it doesn't mean it anything.
I suggest to leave it and maybe add a warning message .
some thing like 
"signal strength units not specified, scaled to 0-100"

arnout replied:
Good idea.

itayx replied:
Added the warning as a comment
and removed the unneeded static casts, but retained the explicit cast to `int32_t` using a functional cast

arnout commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -382,9 +382,13 @@ static bool translate_nl_data_to_bwl_results(sChannelScanResults &results,
-        results.signal_strength_dBm = (nla_get_u32(bss[NL80211_BSS_SIGNAL_MBM])) / 100;
+        //signal strength of probe response/beacon in mBm (100 * dBm) <s32>
+        results.signal_strength_dBm =
+            static_cast<int32_t>(nla_get_u32(bss[NL80211_BSS_SIGNAL_MBM])) / 100;
```
I didn't know static cast could do that! Excellent.

arnout commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -555,40 +555,40 @@ struct BML_STATS_ITER {
+    char ap_SSID[BML_SSID_MAX_LENGTH];
     //The BSSID used for the neighboring WiFi SSID.
-    uint8_t ap_Mode;
+    uint8_t ap_BSSID[BML_MAC_ADDR_LEN];
```
It would have been nicer if the two changes here were in separate commits, because now the unsigned char -> uint8_t change is a bit lost between the rest.

Also, while changing the comments, you could also have converted them to doxygen (just add an additional `/`, so `///`)

Both nitpicks.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2078,3 +2067,25 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+        }
+    }
+    return ss.str();
+}
```
Please implement with a template instead of copying the same code 3 times.

itayx replied:
Changed to use template

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2078,3 +2067,25 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+    std::stringstream ss;
+    if (arr) {
+        for (size_t i = 0; i < arr_max_size; i++) {
+            ss << ((i != 0) ? ", " : "") << static_cast<int>(arr[i]);
```
In this case, we want `int(arr[i])` instead of static_cast, I think. Or better yet, `unsigned int(arr[i])`.

itayx replied:
Changed

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2038,22 +2038,22 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+                          << "  operating_standards=" << int(res.ap_OperatingStandards) << std::endl
+                          << "  operating_channel_bandwidth="
+                          << int(res.ap_OperatingChannelBandwidth) << std::endl
+                          << "  beacon_period_ms=" << (int)res.ap_BeaconPeriod << std::endl
```
You could have converted the casts into initializers as well (nitpick).

itayx replied:
changed to functional cast instead

arnout reviewed (commented):
I have a few important comments that warrant a re-review. However, I might not do the re-review, so just commenting rather than changes requested.

arnout commented:
> Please open an issue for the TLVF bug, and update the commit message to reflect that this change is a workaround but is harmless since uint8 serves the needs just fine.

tlvVarList.yaml actually has this:
```
# TODO tlvf currently doesn't support list length field of more than 1 byte
```

But apparently there's no issue for it yet.

kantera800 commented:
> > Please open an issue for the TLVF bug, and update the commit message to reflect that this change is a workaround but is harmless since uint8 serves the needs just fine.
> 
> tlvVarList.yaml actually has this:
> 
> ```
> # TODO tlvf currently doesn't support list length field of more than 1 byte
> ```
> 
> But apparently there's no issue for it yet.

https://github.com/prplfoundation/prplMesh/issues/836

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.h:
```diff
@@ -205,6 +205,8 @@ class cli_bml : public cli {
                               const std::string &channel_pool);
     int get_dcs_scan_results(const std::string &radio_mac, uint32_t max_results_size,
                              bool is_single_scan = false);
+    template <typename T>
+    const std::string string_from_int_array(T *arr, size_t arr_max_size);
```
Why does it need to be exported now? in other words, why is it extern instead of static and why is it declared in the header file? It is not explained in the commit message.

arnout reviewed (approved):
Looks good, but needs a manual rebase on master.
